org.alloytools.alloy.application/src/main/java/templates/TestTemplateFile.vm

```diff
@@ -412,21 +469,61 @@ class CentralController:
 
     def __init__(self):
         self.concState = $!{rootState.getName()}_State(None, None)
-        self.commands = {"help": self.printHelp, "states": self.printStates, "execute": self.execute}
+        self.concState.default_initialize()
+        self.commands = {"help": self.printHelp, "execute": self.execute}
         #if(${allEnvEvents.size()}> 0)
         self.possibleEvents = {"${allEnvEvents.get(0).getModifiedName()}"#foreach($event in ${allEnvEvents.subList(1, ${allEncEvents.size()})}), "$!{event.getModifiedName()}"#end}
         #else
         self.possibleEvents = {}
         #end
-
+    
+    def handle_event(self):
+        unused_conc_states = self.concState.get_conc_descendants()
+        old_active_states = dict()
+        for conc_state in unused_conc_states:
+            old_active_states[conc_state.__repr__()] = conc_state.get_active_child_state_repr()
+        prev_num_unused_conc_states = -1
+        
+        while len(unused_conc_states) > 0 and not prev_num_unused_conc_states == len(unused_conc_states):
+            prev_num_unused_conc_states = len(unused_conc_states)
+            state_queue = [self.concState]
+            while state_queue:
+                cur_state = state_queue[0]
+                del state_queue[0]
+
+                possible_transitions = []
+                for k, v in cur_state.transitions.items():
+                    if k():
+                        possible_transitions.append(v)
+
+                if possible_transitions:
+                    chosen_transition = random.randint(0, len(possible_transitions) - 1)
+                    possible_transitions[chosen_transition]()
+                    # check if any concurrent states have taken a transition. If so,
+                    # remove concurrent states from the set to prevent additional transitions (big step maximality)
+                    for conc_state in unused_conc_states:
+                        if not conc_state.get_active_child_state_repr() == old_active_states[conc_state.__repr__()]: 
```

Comment:
Don't understand the line here, can you explain?

Reply:
Also `if conc_state.get_active_child_state_repr() != old_active_states[conc_state.__repr__()]:`

Reply:
I think it means "if conc_state.active_child_state is changed". It took me a while to understand it.